clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -0,0 +1,168 @@
+	/subsystem=ejb3/service=timer-service:write-attribute(name=default-persistent-timer-management, value=persistent)
+
+=== Importing Existing Configuration
+
+N/A
+
+=== Deployments
+
+N/A
+
+=== Interoperability
+
+N/A
+
+== Implementation Plan
+
+The feature requires the distributable-ejb subsystem (WFLY-14953) in order to house the configuration for distributed timer management providers.
+
+The general implementation plan for this feature is as follows:
+
+* Introduce a proper SPI for a `TimerService` within `org.jboss.as.ejb3.timerservice.spi`, to facilitate multiple implementations.
+** Adapt the existing `TimerResource` runtime attributes/operations to use the `TimerService` SPI instead of referencing specific implementation classes.
+** Adapt existing `TimerService` implementation to this SPI.
+* Create `CompositeTimerService`, a `TimerService` implementation that delegates to separate persistent and transient `TimerService` implementations.
+* Implement an distributed implementation of the `TimerService` SPI.
+** Add a `org.wildfly.clustering.ejb.timer` SPI to the `wildfly-clustering-ejb-spi` module to facilitate multiple distributed timer manager implementations.
+** Create `DistributedTimerService`, a distributed `TimerService` implementation to the ejb3 subsystem that delegates to the `wildfly-clustering-ejb-spi` for timer management and scheduling.
+** Add an Infinispan implementation of the `org.wildfly.clustering.ejb.timer` SPI that assigns execution ownership of timers to cluster members based on consistent hashing of the timer identifier.
+*** Leverage the existing `LocalScheduler` implementation within `wildfly-clustering-ee-cache` for timer scheduling
+*** Leverage the existing `PrimaryOwnerScheduler` decorator that uses consistent hashing of the timer identifier to distribute timer execution among cluster members
+* Add a management resource and capability to the distributable-ejb subsystem to configure the Infinispan timer management implementation.
+* Make the `default-data-store` attribute of the timer service resource optional.
+* Add an `default-persistent-timer-management` attribute to the timer service resource, as an alternative to `default-data-store`.
+** Require the named distributable timer management capability when defined.
+** Modify the `TimerService` DUP to use a `CompositeTimerService` in the event that the server is configured to use persistent timer management, that delegates to the existing `TimerServiceImpl` for use with transient timers and the `DistributableTimerService` for use with persistent timers.
+
+== Security Considerations
+
+N/A
+
+== Test Plan
+
+* All existing `TimerService` tests must pass against the new persistent timer implementation, including the Jakarta Enterprise Beans TCK.
+* New tests will be added to the clustering integration testsuite to validate correctness of behavior of persistent timers with a cluster, including:
```

Comment:
> Looking at the feature PR it seems that there are no domain tests, correct?

What do you mean by "domain tests"?

> it seems to me that there aren't any tests to ensure that the server can start correctly with the two - mutually exclusive - configuration, except the integration ones, which validate what you detail in the following points.
> Shouldn't such a "smoke" test be part of the PR changes?

In general, it is not feasible to test against every combination of galleon layers - and such testing is truly orthogonal to this feature.
We know that the ejb + ejb-dist-cache layer creates extra distributable-ejb and infinispan cache configuration, but this configuration is not actually mutually exclusive - but it would be a source of confusion for users.  We should definitely address this separately.

Reply:
> What do you mean by "domain tests"?

I mean tests that use the "domain" mode,, not the "standalone" one.

> In general, it is not feasible to test against every combination of galleon layers - and such testing is truly orthogonal to this feature.

I am not asking about Galleon layers, actually. I am asking for a smoke test that validates that the server can start either with the default configuration - which will use the database persistence IIUC - and the new one introduced by the feature. Basically what you describe here:

```
To enable use of the distributable timer service, a user will be required to modify the default timer-service configuration of the ejb3 subsystem as follows:

/subsystem=ejb3/service=timer-service:undefine-attribute(name=default-data-store)
/subsystem=ejb3/service=timer-service:write-attribute(name=default-persistent-timer-management, value=persistent)
```

> We know that the ejb + ejb-dist-cache layer creates extra distributable-ejb and infinispan cache configuration, but this configuration is not actually mutually exclusive - but it would be a source of confusion for users. We should definitely address this separately.

I agree, IMHO there could be room for an issue tracker to improve the documentation. WDYT?

Reply:
> I mean tests that use the "domain" mode, not the "standalone" one.

Domain management is a completely separate function - and is tested separately from any functionality of a host controller.

> I am asking for a smoke test that validates that the server can start either with the default configuration - which will use the database persistence IIUC - and the new one introduced by the feature. Basically what you describe here:

Not currently, though now that we have 2 versions of the distributed timer service test (one using a distributed cache and another using a remote infinispan server), I can modify one variant to change the default behavior, while the other overrides the timer service via deployment descriptor.
The default configuration uses file persistence for persistent timers, BTW.

> I agree, IMHO there could be room for an issue tracker to improve the documentation. WDYT?

Better to fix the layers such that this combination does not include the ejb-local-cache layer.

Reply:
> I am asking for a smoke test that validates that the server can start either with the default configuration - which will use the database persistence IIUC - and the new one introduced by the feature.

OK - I've updated the test that uses a remote infinispan server to configure the distributed timer service via the ejb3 subsystem, rather than using a deployment descriptor.

Reply:
> Domain management is a completely separate function - and is tested separately from any functionality of a host controller.

IIUC, in the WildFly test suite there is room for "smoke"/"basic" tests to verify that a feature configuration is available, in a managed domain, see https://github.com/wildfly/wildfly/blob/main/testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/suites/DatasourceTestCase.java#L44

I understand that not every feature can be tested this way, but in this very case I am wondering whether it would be an opportunity that we could/should consider, given that - with this new feature - there are quite some combinations for managing timers (persistent/transient, on file, database, embedded or "remote" cache etc.) and maybe such configuration options should be tested to be available from a domain controller.
WDYT?

>> ... I am asking for a smoke test that validates that the server can start either with the default configuration - which will use the database persistence IIUC - and the new one introduced by the feature
> Not currently, though now that we have 2 versions of the distributed timer service test (one using a distributed cache and another using a remote infinispan server), I can modify one variant to change the default behavior, while the other overrides the timer service via deployment descriptor.

Looking at the latest changes that you pushed here. IIUC now the HotRod (i.e. remote Infinispan) test is enabling the feature at the subsystem level, thanks to `TimerServiceSetupTask`. 
In such a case, this _integration_ test is the only "smoke" test that we have at the moment to verify that the server starts with the proposed configuration for enabling the feature. Correct? If this is true then a test execution will have to reach integration testing before assessing a potential failuhre in the configuration, right?

> The default configuration uses file persistence for persistent timers, BTW.

Thanks for clarifying, IMHO the community docs could be improved to list the EJB timer service configuration and managment options, for users to easily map their use cases to them.

> Better to fix the layers such that this combination does not include the ejb-local-cache layer.

I'd like to hear from @jfdenise about this, since I am not sure such an issue exists and whether it would be fixed at all, given the relationships between the layers. Will the ejb + ejb-dist combination fixed in a way that no ejb-local-cache will be included implicitly?

Reply:
> IIUC, in the WildFly test suite there is room for "smoke"/"basic" tests to verify that a feature configuration is available in a managed domain

That is not the purpose of the domain testsuite. It is impossible for a feature to *not* be available in a managed domain (unless the management operation explicitly forbids it, which would be very abnormal).

> Looking at the latest changes that you pushed here. IIUC now the HotRod (i.e. remote Infinispan) test is enabling the feature at the subsystem level, thanks to TimerServiceSetupTask.
In such a case, this integration test is the only "smoke" test that we have at the moment to verify that the server starts with the proposed configuration for enabling the feature. Correct? 

Not really.
The ejb3 subsystem unit test validates the use of timer-management references to the distributable-ejb3 subsystem.
The distributable-ejb subsystem unit tests validate the use of cache references to the Infinispan subsystem.
The infinispan subsystem unit tests validate the use hotrod-store references to remote-cache-container resources.

> IMHO the community docs could be improved to list the EJB timer service configuration and managment options, for users to easily map their use cases to them.

Feel free to file a jira to improve the general documentation of the EJB timer service.  This feature is only concerned about documenting the behavior that it adds.

> I'd like to hear from @jfdenise about this, since I am not sure such an issue exists and whether it would be fixed at all, given the relationships between the layers. Will the ejb + ejb-dist combination fixed in a way that no ejb-local-cache will be included implicitly?

I did not get around to filing this bug until now.  Please redirect any discussion of this bug here:  https://issues.redhat.com/browse/WFLY-16326

Reply:
@fabiobrz , my understanding of local vs dist is that it is by design. The local cache is the default and must be excluded to use the dist one.

Reply:
Thanks @jfdenise - I'll report this to https://issues.redhat.com/browse/WFLY-16326, as suggested by @pferraro